doc/index.rst

```diff
@@ -161,70 +164,152 @@ Required environment variables:
+* **TARGET_WORKSPACE** (default: ``$TARGET_REPO_PATH``): Definition of sources for target `workspace <#workspace-management>`__.
+* **UPSTREAM_WORKSPACE** (default: not set): Definition of upstream `workspace <#workspace-management>`__.
 * **VERBOSE_OUTPUT** (default: not set): If `true`, build tool (e.g. Catkin) output prints in verbose mode.
 * **VERBOSE_TESTS** (default: true): If `true`, build tool (e.g. Catkin) output prints in verbose mode during `run_tests` step.
 
-Note: You see some `*PKGS*` variables. These make things very flexible but in normal usecases you don't need to be bothered with them - just keep them blank.
+
+Workspace management
+--------------------
+
+Workflow
+++++++++
+The default test will just build the packages in the target repository and optionally run the contained tests.
+This behavior can be expanded with addtional workspaces
+
+A. Upstream workspace: Source packages that are needed for building or testing the target or downstream packages
+
+   1. Fetch source code (``UPSTREAM_WORKSPACE``)
+   2. Install dependencies with ``rosdep``
+   3. Build workspace ``~/upstream_ws``, chained to /opt/ros
+
+B. Target workspace: Packages in your target repository that should get build and tested
+
+   1. Fetch source code (``TARGET_WORKSPACE``)
+   2. Install dependencies with ``rosdep``
+   3. Build workspace ``~/target_ws``, chained to upstream workspace or /opt/ros
+   4. run tests (opt-out with ``NOT_TEST_BUILD``)
+
+C. Target workspace: Packages that should get tested against your target repository
+
+   1. Fetch source code (``DOWNSTREAM_WORKSPACE``)
+   2. Install dependencies with rosdep
+   3. Build workspace ``~/downstream_ws``, chained to target workspace
+   4. run tests (opt-out with ``NOT_TEST_DOWNSTREAM``)
+
+Workspace definition
+++++++++++++++++++++
+
+Each workspace can be composed as a sequence of the folling items:
+
+* URL of a source repository with the pattern ``<scheme>:<resource>#<version>``, e.g. ``github:ros-industrial/industrial_ci#master``.
+  Supported scheme are:
+
+  * ``github`` for GitHub repositories
+  * ``gitlab`` for Gitlab repositories
+  * ``bitbucket`` for Bitbucket repositories
+  * ``git``/``git+*``: for any other git repository
+
+  For convenience ``@`` (pip-stylecan be used as a separator as well.
+
+* URL (=starts with http or https) of a ``.repos`` or ``*.rosinstall`` file
+* relative path of a ``.repos`` or ``*.rosinstall`` file
+* (relative) directory path to a source directory
+* directory path prefixed with ``-`` to remove the directory
+
+For backwards compatibility, ``UPSTREAM_WORKSPACE`` can be set to ``debian`` and ``file`` as well, but not in combination with the other options and with a deprecation warning.
+In case of ``file``, it will be replaced by ``$ROSINSTALL_FILENAME`` or ``$ROSINSTALL_FILENAME.$ROS_DISTRO``, if the latter exists.
+**In ``file`` mode the target repository will not get removed automatically anymore and therefore might get built twice!**
```

Comment:
I don't understand what you mean with this. Would you mind elaborating a bit?

Reply:
https://github.com/ros-industrial/industrial_ci/blob/master/industrial_ci/src/tests/source_tests.sh#L128 was removed
Without this feature the legacy version would end up in having two packages with the same name -> error.

In the new version these two packages would end up in different workspaces -> no error, but packages in target repositories get build twice.

Reply:
I wasn't familiar with that feature, since we are used to include `.rosinstall` files per-repository which, obviously, do not include the target repository. I guess that makes sense in configurations that point to a common `rosinstall` for many repositories.

I get the point of the comment now, thanks!